### ISSUE TITLE
doc-build should not build documentation with warnings.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6980]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)| @adiroiban reported|
|-|-|
|Trac ID|trac#6980|
|Type|enhancement|
|Created|2014-02-20 10:45:08Z|
|Branch|https://github.com/twisted/twisted/tree/simple-doc-build-fail-6980-2|

If we allow warnings in documentation, there is the risk of accumulating to many warnings and any newly introduced warning will be ignored and not tried to be fixed.

As a new contributed I find it hard to compare old warnings with latest warnings and see if I introduced a warning.

Maybe this is pedantic.. so feel free to ignore this ticket.

Attachments:

* [6980-1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8fb/8fb8d133a0e42e2b5fa401d5801bc25950fe710b/85f906a81835e149a3a7481bbdf738fe4aa1801f.diff) (1721 bytes) - added by adiroiban on 2014-02-20 11:57:38Z - 
* [6980-2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8fb/8fb8d133a0e42e2b5fa401d5801bc25950fe710b/c0a61a6a1b72c8bc74e4b58d460e7eaff110d06f.diff) (1988 bytes) - added by adiroiban on 2014-03-19 09:28:41Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6980 6980
type__enhancement enhancement
reporter__adiroiban adiroiban
priority__normal normal
milestone__ 
branch__branches_simple_doc_build_fail_6980_2 branches/simple-doc-build-fail-6980-2
branch_author__adiroiban__glyph adiroiban, glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1392893108000000 1392893108000000
changetime__1447404551824164 1447404551824164
version__None None
owner__glyph glyph

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_1"></a>|@adiroiban removed owner|
|-|-|

I have attached a patch for this.

It updates the comment in doc-build to provides a clue about how to use the command.
Maybe it is not required.

I have updated the _release.py code to quit when warnings are encoutered.

Sphinx is now run with -q mode to reduce the noise.

I think that for debuging/development devs can use the docs/Makefile.

This is somehow related to [#6901](https://github.com/twisted/twisted/issues/6901) where it is asked that build-doc should give more details.


I agree with exarkun in that the tools designed for build system should output minimal info, but maybe runCommand needs more love.


With the new command, errors are no longer hard to detect.

```
Traceback (most recent call last):
  File "./bin/admin/build-docs", line 29, in <module>
    SphinxBuilder().main(sys.argv[1:])
  File "/home/adi/chevah/twisted/twisted/python/_release.py", line 718, in main
    self.build(FilePath(args[0]).child("docs"))
  File "/home/adi/chevah/twisted/twisted/python/_release.py", line 748, in build
    buildDir.path])
  File "/home/adi/chevah/twisted/twisted/python/_release.py", line 54, in runCommand
    raise CommandFailed(exitCode, None, stdout)
twisted.python._release.CommandFailed: (1, None, "\nWarning, treated as error:\n/home/adi/chevah/twisted/docs/projects/web/howto/web-in-60/access-logging.rst:30: WARNING: Pygments lexer name u'shell' is not known\n\n")
```

but with a bit of love, it can look like this

```
Traceback (most recent call last):
  File "./bin/admin/build-docs", line 29, in <module>
    SphinxBuilder().main(sys.argv[1:])
  File "/home/adi/chevah/twisted/twisted/python/_release.py", line 721, in main
    self.build(FilePath(args[0]).child("docs"))
  File "/home/adi/chevah/twisted/twisted/python/_release.py", line 751, in build
    buildDir.path])
  File "/home/adi/chevah/twisted/twisted/python/_release.py", line 54, in runCommand
    raise CommandFailed(exitCode, None, stdout)
twisted.python._release.CommandFailed: Command failed!
Exit code: 1
Exit signal:None

Warning, treated as error:
/home/adi/chevah/twisted/docs/projects/web/howto/web-in-60/access-logging.rst:30: WARNING: Pygments lexer name u'shell' is not known



```

Following the scout rule, I have updated __str__ method for run command to show information in a nicer format.

Live diff is here https://github.com/chevah/twisted/compare/6980-doc-build-spinx-warnings.diff

It depends on [#6970](https://github.com/twisted/twisted/issues/6970) ... which is a single line diff... 

I will update a patch

**Comment by twisted-trac** [comment]
|[<img alt="multani's avatar" src="https://avatars.githubusercontent.com/u/65311?s=50" width="50" height="50">](https://github.com/multani)<a name="note_2"></a>|@multani commented|
|-|-|

Thanks for the patch!

I check the following:

 * the patch applies cleanly on trunk
 * with the error from [#6970](https://github.com/twisted/twisted/issues/6970), launching `bin/admin/build-docs` now stops the building process with a clear error message
 * Things to do:
   * I think you will need a topfile to get this merged into trunk.

Once you added the missing topfile, I think this is good to be merged. (I don't have commit privileges, you'll need to ask someone who has.)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_3"></a>|@tomprince commented|
|-|-|

I certainly think that it should be easy to see when new warnings are introduced. But

 1. I don't think printing a traceback is the proper way to report errors. Having the build fail is an easy to anticipate error, and the traceback doesn't give any useful information.
 2. I don't think we should turn warnings into errors. Doing so causes the build to be aborted at the first warning, rather than having all the documentation built and warnings shown. 

I think something like a `--verbose` flag suggested in [#6970](https://github.com/twisted/twisted/issues/6970), that displays all the output (perhaps directly to the terminal) is a better solution. With that, buildbot can be made to fail a build when there are any warnings. And, to avoid extraneous output when doing releases, the default could be to pass -q or -Q to sphinx-build.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban commented|
|-|-|

Thank you all for the review.

This patch was submitted thinking that bin/admin tools are only designed for buildbot.

The traceback is generated by runCommand and in case documentation right now contains error and not warning, a similar traceback is reported.

My main concern was to have the command exit with a non-zero code to trigger a buildbot failure.

I tried to keep things simple and reuse the available sphinx commands without adding extra logic for buildbot.

I don't know how to make sphinx exit with non zero code when warnings are present.

I don't want to get very deep into sphinx customization, but maybe an upstream bug report could help.

For the time being, I would like to enforce a clean documentation build as soon as possible.

I don't know how to make sphinx command fail when there are warning. This is why I went for turning warnings into errors. I am open to suggestions.

My workflow is that once a developer observe a doc failure on buildbot, he/she will run doc/Makefile on his/her own computer and observer all errors and fix then. If previous build was clean, then we know for sure that all error are introduced by him/her.

I agree that we can improve the tools and that improvement can be done as part of [#6970](https://github.com/twisted/twisted/issues/6970) .. Including traceback reporting.


Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_5"></a>|@tomprince commented|
|-|-|

So ... buildbot can be made smart enough to fail when there are sphinx warnings, even if the command doesn't exit non-zero *if* the output of the command is the usual sphinx output. (It will even extract the warnings and errors to display them separately).

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban commented|
|-|-|

I prefer to have less logic included only in buildot and developers to run all commands on local computers, without buildbot as a dependency.

I prefer to have smarted bin/admin tools and dumber buildbot.

What do you think? Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_7"></a>|@tomprince commented|
|-|-|

> I prefer to have smarted bin/admin tools and dumber buildbot. 

Sure. But I don't think failing on the first warning and not generating output is smarter.

 1. It makes fixing all the warnings harder. Sure, they could run another tool, to do that (but again, that doesn't seem smarter).
 2. Seeing a later warning might help in diagnosing an earlier one. (For example, if there are 20 instances of an warning, the way to approach it might be different than if there was only a single warning).
 3. Seeing the generated output might make it obvious what the warning is.

Also, I don't about most people, but I'm more likely to notice output, than the exit value of programs that I run (even if I do have non-zero exit values displayed in my shell). So buildbot looking at the output is exactly what a developer would likely do.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph set owner to @adiroiban|
|-|-|

HI adiroiban,

Thanks for your contribution.

I agree with pretty much everything Tom has said here.  We should not be failing on the first error, and not producing output is a separate consideration than failing when there are warnings.

The new code in `_release` also has no test modifications, and the new untested `__str__` decreases code coverage statistics as well.

However, I definitely agree that this logic should be available locally for developers and should not necessitate going out to the buildbot.

Please submit a version of this that addresses tom.prince's feedback.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban commented|
|-|-|

(In [46058]) Branching to doc-build-fail-6980.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_10"></a>|@adiroiban removed owner|
|-|-|

Thanks for your feedback!

I don't understand why are you against the previous patch.

I understand the design for the perfect build tool.
I was not trying to create the perfect build tool, but just to make it a little better.

The current output of bin/admin/build-docs is just nothing... so failing
on first error has nothing old to hide.

https://buildbot.twistedmatrix.com/builders/documentation/builds/6114/steps/process-docs/logs/stdio

---------

Anyway, I am still trying to make the current build a little better :)

I have pushed a new version which writes the errors and warnings into a log file.

With this we should be able to configure the documentation builder to fail if the log file is not empty.

To test the log file:

```
$ ./bin/admin/build-docs . err.log
$ ./bin/admin/build-docs .
```

I have only written test for the SphinxBuilder.main() method... I think that the build() method should be made private.

Right now we have a few warnings in docs/core/howto/logger.rst ... and if we don't update the documentation builder, I guess that we will accumulate more.

Please check the branch and suggest changes.

Many thanks for your feedback!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph set owner to @glyph|
|-|-|

Replying to [adiroiban](#note_10):
> Thanks for your feedback!
> 
> I don't understand why are you against the previous patch.

It's been a while, I'm not sure I remember either.  I just thought Tom's feedback was valid, and things seemed stuck, so I converted it to an official review.

> I understand the design for the perfect build tool.
> I was not trying to create the perfect build tool, but just to make it a little better.

Point taken.

> The current output of bin/admin/build-docs is just nothing... so failing
> on first error has nothing old to hide.
> 
> https://buildbot.twistedmatrix.com/builders/documentation/builds/6114/steps/process-docs/logs/stdio

Oh, um.  I did not realize that.  I guess I have been often invoking sphinx directly and expecting to see its output.

> I have pushed a new version which writes the errors and warnings into a log file.

What we really want to happen here is to get an error code from build-docs, so that we can handle that as normal, and consider it a build failure.  Having to special-case a log file on our build infrastructure is yet more special configuration, and makes it hard to make i.e. `tox` handle warnings as errors correctly.

There's a completely different strategy which will achieve the desired result, so I am going to try doing that in a new branch; you can review, and if you don't like it, put this branch back up :).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph commented|
|-|-|

(In [46169]) Branching to simple-doc-build-fail-6980-2.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph removed owner|
|-|-|

New branch now building; should be ready for review.  Hopefully the documentation builder will just fail with no need for further configuration; if so, I'll go ahead and fix the logger documentation errors too.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph commented|
|-|-|

Sweet.  Build failed: https://buildbot.twistedmatrix.com/builders/documentation/builds/6162/steps/process-docs/logs/stdio

I'll fix the warnings since there are so few.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_15"></a>|@glyph commented|
|-|-|

When the builders finish running, https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/simple-doc-build-fail-6980-2&num_builds=2 should show a failed doc build and then a successful one.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_16"></a>|@adiroiban set owner to @glyph|
|-|-|

Changes looks great. Thanks.

Happy to see this fixed :) and improve the tools to help us continue to keep the code clean.

Please merge.

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph set status to `closed`|
|-|-|

(In [46174]) Merge simple-doc-build-fail-6980-2: fail on sphinx

Author: glyph

Reviewer: adiroiban

Fixes: [#6980](https://github.com/twisted/twisted/issues/6980)

Sphinx warnings will now cause build-docs to fail, since we should not
have warnings in our docs.

This change also therefore fixes our few lingering sphinx warnings, so
the buildbot will not fail when it lands.
